state: Add terminating gateway events for streaming

dnephin opened the pull request:
TODO:
* [x] add a terminating gateway to snapshot test
* [x] review the comparison operation to make sure it is not hiding bugs (the comment below)
* [x] include enterprise meta changes

dnephin commented on agent/consul/state/catalog_events_test.go:
```diff
@@ -895,7 +1484,8 @@ var cmpPartialOrderEvents = cmp.Options{
 		key := func(e stream.Event) string {
 			csn := getPayloadCheckServiceNode(e.Payload)
-			return fmt.Sprintf("%s/%s/%s", e.Topic, csn.Node.Node, csn.Service.Service)
+			// TODO: double check this sort key is correct.
+			return fmt.Sprintf("%s/%s/%s/%s", e.Topic, csn.Node.Node, csn.Service.Service, e.Payload.(EventPayloadCheckServiceNode).key)
```
Leaving a comment to call out a TODO, so we double check this is safe and won't hit any problems.

There may be a safer way to do this by using `MatchesKey`

banks reviewed (approved):
LGTM! 

I'm not sure if you addressed the remaining TODO as the comment was on an old commit and is now "outdated". The one thing that occurs to me is that in Ent we would need namespace as part of that key presumably and I saw that at least some of these tests have a version that pulls in extra cases for Ent while using the same test func so that might matter then?

dnephin commented:
I rebased with master to prevent any logical conflicts on merge.

The sort order seemed fine, but I changed the logic a bit to only use the overideKey or the service name, but not both, to match MatchesKey. Also added namespace to the key.
